### DISCUSSION TITLE
Gosund Nite Bird LB3 add new support NEED HELP

### DISCUSSION THREAD

**Discussion by pepsonEL** [arendst/Tasmota#10692] [Q&A] [answered]
I but this light and it has ESP8266. I flashed it by Tasmota 9.2 but i have a problem with setup GPIO for PWM. It has 5 colors led: RGB and Cold White and Warm WHite. Is anybody who can help me step by step how setup it  and it can support for all new devices. This light has on front buuton to change betwen white and colors and two slider on left and right to setup brightnes. Double tak on top board turn on and off light. It is image motheborad.
![IMG_20210124_222255](https://user-images.githubusercontent.com/71824185/105744588-a58c2100-5f3d-11eb-95af-2ad7f8de78bb.jpg)
![IMG_20210124_222419](https://user-images.githubusercontent.com/71824185/105744622-ae7cf280-5f3d-11eb-9f0e-0164fb07a800.jpg)

![Uploading IMG_20210124_222515.jpg…]()
![IMG_20210124_223022](https://user-images.githubusercontent.com/71824185/105744664-ba68b480-5f3d-11eb-916d-8ec03489ad7c.jpg)
![Uploading IMG_20210124_223108.jpg…]()

**Comment by sfromis** [comment, marked as answer]
https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/
After having found what gpios are for PWM, `SetOption68` makes it easier to get the details right.

**Comment by pepsonEL** [comment]
Can you help me more and explaine how use SetOption68 please...

**Reply by ascillato2** [reply]
To write a command in Tasmota, you have to open the Tasmota WebUI and click on console and type there the commands.

The full explanation on how using commands is in the docs at https://tasmota.github.io/docs/Commands

Then, at https://tasmota.github.io/docs/Commands/#setoptions you have what `SetOption68` does

![image](https://user-images.githubusercontent.com/39969427/105760032-f39a2800-5f2f-11eb-94e9-f8b83348cbca.png)

---------------------------

### Support Information 
See [Docs](https://tasmota.github.io/docs/) for more information.
See [Chat](https://discord.gg/Ks2Kzd4) for more user experience.
See [Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)

**Comment by pepsonEL** [comment]
Ok i set this:
Backlog SetOption0 0; SetOption36 1

then set for all GPIO relay as show:
![relay](https://user-images.githubusercontent.com/71824185/105768076-265a1580-5f5c-11eb-9692-e6f5b1d0e8dc.jpg)

and on GPIO5 relay 4 -- light color BLUE
GPIO12 relay 5 -- light color GREEN
GPIO13 relay 6 --light color white(probably warm white)
GPIO15 relay 8 --light color RED

on other GPIO not light. WHat next i must do ?

**Comment by pepsonEL** [comment]
ok setup as: 
GPIO15 - R - PWM1
GPIO12 - G - PWM2
GPIO5 - B - PWM3
GPIO13 - WW - PWM4

works ok. 
But next step.... How add support to touch button. Double tap/touch  on top case should TURN ON and OFF LIGHT. 
Touch center button fron should switch between colors light and white light.
And slider touch on right and left should change brightnes.

**Comment by pepsonEL** [comment]
ANybody can help me please....

**Reply by barbudor** [reply]
Did you follow the procedure and identified the GPIO for the buttons ?

If the touch slider are the set of little dots we can see on the right of your 1st and 3rd photos (close to number "2"), there is apparently a chip dedicated for managing the touch buttons
You need to identify what is the chip from its marking and find the datasheet of it to understand how it is working and how it could be communicating with the ESP.
Try to find connections between this chip and the ESP. Could be serial for example.

By purchasing a device that was not on the template repository, you took the risk of having to find out everything by yourself or even finding no solutions at all.

**Comment by pepsonEL** [comment]
No it is not touch button. Buttons are on the cover and is connected to thos two connector about numer "2" on motheborad.
GPIO i try to button but not working.

**Comment by pepsonEL** [comment]
![IMG_20210127_154905](https://user-images.githubusercontent.com/71824185/106009088-46065080-60b8-11eb-96ec-6ede205af234.jpg)
![IMG_20210127_154910](https://user-images.githubusercontent.com/71824185/106009112-4b639b00-60b8-11eb-9873-7f8f0a94f805.jpg)
![IMG_20210127_154918](https://user-images.githubusercontent.com/71824185/106009124-4d2d5e80-60b8-11eb-989f-87ea195942bb.jpg)

**Comment by barbudor** [comment]
Still the signals from the side connectors seems to go to the dots then resistors then the big square chip. 
So what is this chip? How is it connected to the ESP? 
We don't have the device in hand so we can't do this analysis for you. You have to do the work.

**Comment by pepsonEL** [comment]
Ok i know.It is chip KINFU CHIPON KF8TS2716
https://www.chipon-ic.com/upload/file/20200506/f25983eb-c255-45d1-8da2-65afcc3bb04f.pdf?fbclid=IwAR1prWNPZxMqL-u5Kg9NR708atvml9jP9K1TBgeW2gd0I4_bwfVb_aCgEiw

 But now i want back to original FW which backup by Tasmotizer and also other by esptool for detect chip 2MB. I done it by command:
python esptool.py -b 115200 --port COM3 read_flash 0x00000 0x200000 flash_2M.bin


But now when i flash it(original) to my light , my light not working. No flash by light nothing. WHen again flash tasmota it works. 
Why on original not working... please help me....

**Reply by Golatas** [reply]
Hi, bought a LB3 today, as well. Disassembled it since tuya-convert did not work. However, did not flash it with tasmota, for now. Maybe I can help you in getting some information from a module with original firmware and getting yours working again, before I flash Tasmota on mine :-)

**Comment by randyg503** [comment]
Curious if any further progress has been made on getting this device added to the Tasmota Device Templates Repository?

**Reply by randyg503** [reply]
I noticed that this device was added to the Tasmota Device Templates Repository on 2022-01-10.  It appears that as of this writing, there is not yet support for the i2c based touch buttons.

**Reply by sfromis** [reply]
From the template page for [Gosund LB3](https://templates.blakadder.com/gosund_LB3):
> Please note, this template currently only allows control of the lights, it does not yet have access to the i2c based touch buttons (anyone want to help?)

**Reply by barbudor** [reply]
The touch is under control of a [KF8TS2716 ](https://www.chipon-ic.com/AboutUs/NewsDetails/1d858902-aeda-48ad-a723-7ce7f1962908) which is a generic MCU with built-in touch-button controller and acting as an I2C slave.
This means that the I2C commands/response are proprietary under control of the manufacturer's firmware inside that MCU and is not documented.
To tackle this down someone with the lamp running the original ESP software needs to hook an I2C analyser to reverse-engineering the protocol.
Only way to move on.

**Reply by bfranske** [reply]
FWIW I was able to serial flash one of these today with Tasmota. Unfortunately I do not have time to work on code for the I2C button/slider system...

@barbudor the good news though is that I was able to find some documentation on the I2C protocol from the Xiomi bslamp2 which uses the same chip: https://github.com/mmakaay/esphome-xiaomi_bslamp2/blob/dev/doc/technical_details.md#front-panel

According to https://github.com/esphome/feature-requests/issues/1096 the I2C GPIOs in use are: GPIO2 - I2C SDA, GPIO14 - I2C SCL

Hopefully this will allow someone to do some work on the I2C support, especially since it seems like it could be reused for other devices using the KF8TS2716

**Reply by sfromis** [reply]
"Same chip" does not give anything when the chip is a generic MCU. Or do you have indications that both lamps are using the same firmware on that chip?
